File: COMPILADOR/kokoslan_/src/src/java/ast/KoKoBiOperation.java
/**
 @author loriacarlos@gmail.com
 
*/
package src.java.ast;
import java.util.*;
import src.java.eval.KoKoContext;
import src.java.eval.KoKoNumValue;
import src.java.eval.KoKoValue;
import src.java.exception.KoKoEvalException;


public class KoKoBiOperation extends KoKoOperation {
	
    public KoKoBiOperation(KoKoAst operator, KoKoAst left, KoKoAst right){
		super(operator, Arrays.asList(left, right));
	}
	
	public KoKoAst left(){return this.operands.get(0);}
	public KoKoAst right(){return this.operands.get(1);}
	
	@Override
	public KoKoValue eval(KoKoContext ctx){
	   try {
	        KoKoId operId = (KoKoId)operator;
			KoKoNumValue lv = (KoKoNumValue)(left().eval(ctx));
			KoKoNumValue rv = (KoKoNumValue)(right().eval(ctx));
			switch (operId.getValue()){
				case "+" : return new KoKoNumValue(lv.getValue() + rv.getValue());
				case "-" : return new KoKoNumValue(lv.getValue() - rv.getValue());
				default  : throw new KoKoEvalException("KoKoBiOperation: unimpemented operator");
			}
			
	   } catch (Exception e) {
			throw new KoKoEvalException(e.getMessage());
	     }
	   
   }
	
}

File: COMPILADOR/kokoslan/src/KoKoKotlin/ast/KoKoBiOperation.kt
/**
 * @author loriacarlos@gmail.com
 */
package KoKoKotlin.ast

import KoKoKotlin.eval.KoKoContext
import KoKoKotlin.eval.KoKoNumValue
import KoKoKotlin.eval.KoKoValue
import KoKoKotlin.exception.KoKoEvalException
import java.util.*

import src.java.ast.KoKoAst



open class KoKoBiOperation(operator: KoKoAst, left: KoKoAst, right: KoKoAst) : KoKoKotlin.ast.KoKoOperation(operator, Arrays.asList(left, right)) {

    fun left(): KoKoAst {
        return this.operands[0]
    }

    fun right(): KoKoAst {
        return this.operands[1]
    }

    override fun eval(ctx: KoKoContext): KoKoValue {
        try {
            val operId = operator as KoKoKotlin.ast.KoKoId
            val lv = left().eval(ctx) as KoKoNumValue
            val rv = right().eval(ctx) as KoKoNumValue
            when (operId.value) {
                "+" -> return KoKoNumValue(lv.value + rv.value)
                "-" -> return KoKoNumValue(lv.value - rv.value)
                else -> throw KoKoEvalException("KoKoBiOperation: unimpemented operator")
            }

        } catch (e: Exception) {
            throw KoKoEvalException(e.message)
        }

    }

}